rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+# Proposal to add a reusable kernel knob
+
+## Introduction
+
+Primitive creation is an expensive operation, which is why oneDNN has a
+primitive cache, a persistent cache, and a kernel cache (for the GPU). For GPU
+implementations, the kernel cache is optional and must be opted into. Although
+compiling kernels for a given problem shape can open up optimization potential,
+it comes at the cost of being incompatible with the kernel cache because the
+kernel cannot be reused for different problem shapes. Therefore, it can be
+helpful to the user to be able to specify if they prefer to use a kernel that
+has been designed to be reusable via the kernel cache (accepting any possible
+performance impacts associated with it) or the typical performance-optimized
+kernel. This proposal is to add a knob the user can control to specify whether
+they would like reusable kernels to be dispatched preferentially over the
+standard performance-optimized kernel ordering for GPU implementations.
```

Comment:
>  Can you provide examples that fundamentally require non-reusable implementations to get high performance? @echeresh, are there any such operations you have noted in convolution?

With enough specialization we can get high enough performance. But the level of specialization is important: e.g. due to opportunities for unrolling and reduced arithmetic we may need to have a specialization for 3x3 kernel. In practice it may not be feasible to specialize for all such cases so we have to trade performance for reusability. I think it's fine to assume reusable kernels can give us high enough performance but full JITting can add a little bit more.

Reply:
@echeresh, You are not quite answering what I am asking. While there is definitely a specialization vs performance tradeoff, an attribute to prefer reusable kernels does not address this. I am more wondering if there are useful specializations we currently used which fundamentally oppose to kernel reuse. For an example, consider the following:
```
int div(int m, int n) {
    return m/n;
}
```

Using our current JIT method, we can implement this operation significantly faster by relying on division by a constant to transform the division into a combination of shifts, adds, and multiplications. I consider this specialization as fundamentally opposed to kernel reuse though, as it requires specialization for every possible input `n`. We actually hit this issue in the reusable bathnorm implementation. The process of transforming `global_ids -> dim ids -> buffer offset`. required division/modulus, so a naive translation had poor performance. This issue was avoided by a more direct calculation of  `global_ids -> buffer_offset`. This calculation is not always efficient due to packing n-dimensions into the 3d range supported by the runtime, but in practice, I suspect this is not an issue. I am wondering if you have seen any other "fundamentally opposed" specializations like this.

Reply:
> Using our current JIT method, we can implement this operation significantly faster by relying on division by a constant to transform the division into a combination of shifts, adds, and multiplications.

For this particular operation it's not that bad:

- Division is performed before the main loop hence it adds only a constant overhead per thread
- Division by a non-constant can be done by pre-computing special "magic" values on the host so we end up with similar instructions for shifting/multiplication. It's not for free as this results in extra overhead but this is not something significant

So I don't think there is something fundamentally opposed to kernel reuse. At the same time full JITting provides opportunities for optimization as we can avoid passing/reading kernel arguments and reduce arithmetic/branching in some cases (like division by power of two, unrolling of loops, etc).

Reply:
Thanks @echeresh. That aligns with my expectation, but I think it is important to confirm it. My personal opinion is that we should not maintain the infrastructure for full specialization unless we find  "fundamentally opposed" specializations that provide significant benefit (maybe defined as >5% performance increase on multiple KPI's). Instead, I think it will be more productive to focus our attention on improving methods for `finding/dispatching/implementing` reusable specializations. While there may be a need for a trade-off controlling attribute, with convolution/gemm the only implementations requiring it, I think nGEN based generation may be fast enough to not justify the extra `finding/dispatching` complexity.

Pulling this back around to the details in this RFC @Simonsays095 , if our eventual goal is to have every implementation be reusable, then this attribute will become obsolete. As such, I wonder if it is justified to include it in the library API. I think we need a more clarity on what the long term goal is, whether this attribute is necessary, or if we should investigate a `trade-off` control attribute more directly, especially as that has appeared in discussions here multiple times.

Reply:
I'm working on updating the RFC to address some confusion about the goal, but the short answer is that we need a near-term solution for frameworks to be able to select reusable kernels preferentially, even if they are not optimized or near performance parity with our JIT kernels. In the short term, feedback from frameworks is that such control over the creation time/execution time tradeoff is needed to optimize performance. Additionally, this will allow us to more rapidly develop reusable kernels without requiring performance parity. In the long term, it's possible that such a control becomes obsolete (after many implementations have been rewritten and updated - likely several years from now), however I would argue it's likely that there will always be cases (whether due to limitations of _existing_ implementations or due to _fundamental_ restrictions on reusable kernels) where this tradeoff can be leveraged to improve performance.

Reply:
> My personal opinion is that we should not maintain the infrastructure for full specialization unless we find "fundamentally opposed" specializations that provide significant benefit

I agree. There are two aspects in the context of moving from one paradigm (JIT) to another (reusable kernels):

- Development resource allocation. From some point going forward we should prefer investing more resources in improving "reusable" kernels rather than JIT kernels
    - For some primitives it may be more straightforward to reach the same level and drop the non-reusable kernels completely (think eltwise), for other primitives - we have to maintain two paths for some time
- Performance gap between JIT vs reusable kernels. It will likely take quite a bit of time to close/reduce the gap so we can address this gradually:
    - Step 1. Expose a control to the user to opt in for the new reusable kernels (this RFC). This way we can continue development of reusable kernels bringing more performance/functional improvements
    - Step 2. Once performance is at a good level and coverage is acceptable we make it default starting from some future architecture